bpf/probe.bpf.c

```diff
@@ -0,0 +1,125 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "utils.h"
+#include "stdbool.h"
+#include "bpf_dbg.h"
+#include "go_types.h"
+#define MAX_BUCKETS 8
+#define W3C_KEY_LENGTH 11
+#define W3C_VAL_LENGTH 55
+
+struct
+{
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(key_size, sizeof(u32));
+    __uint(value_size, sizeof(struct map_bucket));
+    __uint(max_entries, 1);
+} golang_mapbucket_storage_map SEC(".maps");
+
+static __always_inline _Bool bpf_memcmp(char *s1, char *s2, s32 size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        if (s1[i] != s2[i])
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static __always_inline void *extract_context_from_req_headers(void *headers_ptr_ptr)
+{
+    void *headers_ptr;
+    long res;
+    res = bpf_probe_read(&headers_ptr, sizeof(headers_ptr), headers_ptr_ptr);
+    if (res < 0)
+    {
+        return NULL;
+    }
+    u64 headers_count = 0;
+    res = bpf_probe_read(&headers_count, sizeof(headers_count), headers_ptr);
+    if (res < 0)
+    {
+        return NULL;
+    }
+    if (headers_count == 0)
+    {
+        return NULL;
+    }
+    unsigned char log_2_bucket_count;
+    res = bpf_probe_read(&log_2_bucket_count, sizeof(log_2_bucket_count), headers_ptr + 9);
```

Comment:
Can you name these two constants 9 and 16 below? What do they represent?

Reply:
I added the hmap struct from https://go.dev/src/runtime/map.go and used offsetof to make it more clear

Reply:
Ah yeah that's lot more clear now :). One thing to consider is that if we copy code from the Go runtime directly here we need to copy their license information and copyright too. I see couple of options:

1. Copy the copyright and the license from the Go runtime code.
2. Create 2 constants OFFSET_OF_GO_RUNTIME_HMAP_FIELD_...(or whatever is reasonable)
3. Add a comment next to the + constant); saying what that constant is.

I personally prefer that we do option 2., especially since we already have means to find what the actual offsets are for a given version of Go, why wouldn't we use the offsets.json for this? It will make the code more portable in the unlikely event that the hmap implementation in Go changes offsets.

Reply:
I agree, option 2 has been done.